rclrs/src/publisher.rs

```diff
@@ -231,6 +231,38 @@ where
     }
 }
 
+/// `PublisherOptions` are used by [`Node::create_publisher`][1] to initialize
+/// a [`Publisher`].
+///
+/// [1]: crate::Node::create_publisher
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct PublisherOptions<'a> {
+    /// The topic name for the publisher.
+    pub topic: &'a str,
+    /// The quality of service settings for the publisher.
+    pub qos: QoSProfile,
+}
+
+impl<'a> PublisherOptions<'a> {
+    /// Initialize a new [`PublisherOptions`] with default settings.
+    pub fn new(topic: &'a str) -> Self {
+        Self {
+            topic,
+            qos: QoSProfile::topics_default(),
+        }
+    }
+}
+
+impl<'a, T: IntoPrimitiveOptions<'a>> From<T> for PublisherOptions<'a> {
+    fn from(value: T) -> Self {
+        let primitive = value.into_primitive_options();
+        let mut options = Self::new(primitive.name);
+        primitive.apply(&mut options.qos);
```

Comment:
This just kind of feels weird. `PrimitiveOptions` is almost like a prototype `QoSProfiles` with option on every field + a name. If you have the `impl From` for `PrimitiveOptions` -> `QoSProfiles` this could read like

```rust
impl<'a, T: IntoPrimitiveOptions<'a>> From<T> for PublisherOptions<'a> {
    fn from(value: T) -> Self {
        let primitive = value.into_primitive_options();
        Self {
            topic: primitive.name,
            qos: QoSProfile::from(primitive),
        }
    }
}
```

Nice thing here is you shouldn't need to mutate anything

Reply:
> is almost like a prototype QoSProfiles with option on every field + a name.

For the time being, that's the primary role of `PrimitiveOptions`. That role could expand in the future if we have a need for more options that apply across all primitive types.

But there's a specific reason that every field in `PrimtiiveOptions` is `Option`: The role of each field is to **override** each QoS profile setting that the user has explicitly specified, and otherwise just use the default setting **that is specific the the primitive type which is being  created**.

Notice that [`SubscriptionOptions`](https://github.com/mxgrey/ros2_rust/blob/d67f04136da2d7682ff05fc465d6554dd889781d/rclrs/src/subscription.rs#L284) and [`ServiceOptions`](https://github.com/mxgrey/ros2_rust/blob/d67f04136da2d7682ff05fc465d6554dd889781d/rclrs/src/service.rs#L202) are initialized with different QoS settings. We use `fn apply(...)` so that we override those settings after their defaults have been set up according to what type of primitive is being created.

If we did what you're suggesting here then all primitive types would always have the same default QoS settings.